File: core/src/nl/t64/game/rpg/screens/world/PartyMembersLoader.java
package nl.t64.game.rpg.screens.world;

import nl.t64.game.rpg.Utils;
import nl.t64.game.rpg.components.party.HeroItem;
import nl.t64.game.rpg.components.party.PartyContainer;
import nl.t64.game.rpg.screens.world.entity.*;
import nl.t64.game.rpg.screens.world.entity.events.LoadEntityEvent;

import java.util.ArrayList;
import java.util.List;


class PartyMembersLoader {

    private final Entity player;
    private final List<Entity> partyMembers;

    PartyMembersLoader(Entity player) {
        this.player = player;
        this.partyMembers = new ArrayList<>(PartyContainer.MAXIMUM - 1);
    }

    List<Entity> loadPartyMembers() {
        Utils.getGameData().getParty().getAllHeroes()
             .stream()
             .filter(this::isHeroNotPlayer)
             .map(this::createEntity)
             .forEach(this::addToPartyMembers);

        return List.copyOf(partyMembers);
    }

    private boolean isHeroNotPlayer(HeroItem hero) {
        return !hero.isPlayer();
    }

    private Entity createEntity(HeroItem hero) {
        String heroId = hero.getId();
        return new Entity(heroId, new InputPartyMember(), new PhysicsPartyMember(), new GraphicsPartyMember(heroId));
    }

    private void addToPartyMembers(Entity partyMember) {
        partyMembers.add(partyMember);
        partyMember.send(new LoadEntityEvent(EntityState.IDLE, player.getDirection(), player.getPosition().cpy()));
    }

}

File: core/src/nl/t64/game/rpg/screens/world/PartyMembersLoader.kt
package nl.t64.game.rpg.screens.world

import nl.t64.game.rpg.Utils.brokerManager
import nl.t64.game.rpg.Utils.gameData
import nl.t64.game.rpg.components.party.HeroItem
import nl.t64.game.rpg.components.party.PartyContainer
import nl.t64.game.rpg.screens.world.entity.*
import nl.t64.game.rpg.screens.world.entity.events.LoadEntityEvent


internal class PartyMembersLoader(private val player: Entity) {

    private val partyMembers: MutableList<Entity> = ArrayList(PartyContainer.MAXIMUM - 1)

    fun loadPartyMembers(): List<Entity> {
        createPartyMembers()
        return ArrayList(partyMembers)
    }

    private fun createPartyMembers() {
        gameData.party.getAllHeroesAlive()
            .filter { !it.isPlayer }
            .map { createEntity(it) }
            .forEach { addToPartyMembers(it) }
    }

    private fun createEntity(hero: HeroItem): Entity {
        return Entity(hero.id, InputPartyMember(), PhysicsPartyMember(), GraphicsPartyMember(hero.id))
    }

    private fun addToPartyMembers(partyMember: Entity) {
        partyMembers.add(partyMember)
        brokerManager.detectionObservers.addObserver(partyMember)
        partyMember.send(LoadEntityEvent(EntityState.IDLE, player.direction, player.position.cpy()))
    }

}
